crates/wasmlink/runtime/src/lib.rs

```diff
@@ -0,0 +1,191 @@
+            Entry::Empty { .. } => None,
+        }
+    }
+
+    pub fn remove(&mut self, idx: u32) -> Option<T> {
+        let slot = self.storage.get_mut(idx as usize)?;
+        match mem::replace(slot, Entry::Empty { next: self.next }) {
+            Entry::Full(b) => {
+                self.next = idx as usize;
+                Some(b)
+            }
+            Entry::Empty { next } => {
+                *slot = Entry::Empty { next };
+                None
+            }
+        }
+    }
+}
+
+impl<T> Default for Slab<T> {
+    fn default() -> Slab<T> {
+        Slab {
+            storage: Vec::new(),
+            next: 0,
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct IndexSlab {
+    slab: Slab<ResourceIndex>,
+}
+
+impl IndexSlab {
+    pub fn insert(&mut self, resource: ResourceIndex) -> u32 {
+        self.slab.insert(resource)
+    }
+
+    pub fn get(&self, slab_idx: u32) -> Option<ResourceIndex> {
+        self.slab.get(slab_idx).cloned()
+    }
+
+    pub fn remove(&mut self, slab_idx: u32) -> Option<ResourceIndex> {
+        self.slab.remove(slab_idx)
+    }
+}
+
+#[derive(Default)]
+pub struct ResourceSlab {
+    slab: Slab<Resource>,
+}
+
+#[derive(Debug)]
+struct Resource {
+    wasm: u32,
+    references: u32,
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct ResourceIndex(u32);
+
+impl ResourceSlab {
+    pub fn insert(&mut self, wasm: u32) -> ResourceIndex {
+        ResourceIndex(self.slab.insert(Resource {
+            wasm,
+            references: 1,
+        }))
+    }
+
+    pub fn get(&self, idx: ResourceIndex) -> u32 {
+        self.slab.get(idx.0).unwrap().wasm
+    }
+
+    pub fn clone(&mut self, idx: ResourceIndex) {
+        let resource = self.slab.get_mut(idx.0).unwrap();
+        resource.references = resource.references.checked_add(1).unwrap();
+    }
+
+    pub fn drop(&mut self, idx: ResourceIndex) -> Option<u32> {
+        let resource = self.slab.get_mut(idx.0).unwrap();
+        assert!(resource.references > 0);
+        resource.references -= 1;
+        if resource.references != 0 {
+            return None;
+        }
+        let resource = self.slab.remove(idx.0).unwrap();
+        Some(resource.wasm)
+    }
+}
+
+lazy_static::lazy_static! {
+    static ref SLABS: Mutex<Vec<(IndexSlab, ResourceSlab)>> = Mutex::new(Vec::new());
+}
+
+#[no_mangle]
+pub extern "C" fn resource_insert(id: u32, res: u32) -> u32 {
+    let mut slabs = SLABS.lock().unwrap();
+
+    if slabs.len() <= id as usize {
+        slabs.resize_with(id as usize + 1, Default::default);
+    }
+
+    let (indexes, resources) = slabs.get_mut(id as usize).unwrap();
+    let index = resources.insert(res);
+    indexes.insert(index)
+}
+
+#[no_mangle]
+pub extern "C" fn resource_get(id: u32, idx: u32) -> u32 {
+    let mut slabs = SLABS.lock().unwrap();
+
+    if slabs.len() <= id as usize {
+        slabs.resize_with(id as usize + 1, Default::default);
+    }
+
+    let (indexes, resources) = slabs.get(id as usize).unwrap();
```

Comment:
FWIW I think it's fine to have `id: usize` in the signature since it becomes `i32` at the wasm layer anyway